feat: Throw when encountering unexpected RPC method hooks

rekmarks opened the pull request:
## Description

Removes RPC handler hooks that had become unused. In order to prevent this from occurring again, also adds a check to fail immediately on dapp connections if extraneous hooks are provided. Finally, adds tests for `createMethodMiddleware.js`.

#### Note for reviewers

- In an indication that this is useful, you'll note that the set of hooks removed in this PR is different from that the original PR (#20580).
- In the commit history, you'll see an added and reverted TypeScript conversion of the method middleware and its tests. The RPC method handler types are a mess and we should probably wait to convert until we've resolved some upstream issues. Either way, the conversion would be too distracting for this PR.

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

rekmarks committed:
feat: Throw error for extraneous method middleware hooks

rekmarks committed:
fix: Mock method middleware in MetaMaskController test

Fixes unit test failures by mocking the method middleware in
metamask-controller.test.js.

rekmarks committed:
refactor: Remove extraneous RPC handler hooks

legobeat commented on app/scripts/lib/rpc-method-middleware/createMethodMiddleware.js:
```diff
@@ -33,14 +34,23 @@ const expectedHookNames = Array.from(
     );
   }
 
+  const extraneousHookNames = Object.keys(hooks).filter(
```
Would this be more appropriate?
```suggestion
  const extraneousHookNames = Object.getOwnPropertyNames(hooks).filter(
```
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyNames

rekmarks replied:
I don't think we have reason to expect keys on the objects in question to be non-enumerable, but sure, why not! Done.

rekmarks committed:
refactor: Move hook assertions into own function

rekmarks committed:
refactor: Use getOwnPropertyNames() instead of keys()

rekmarks committed:
refactor: Convert createMethodMiddleware to TypeScript

rekmarks committed:
test: Add createMethodMiddleware tests

rekmarks committed:
refactor: Improve method middleware types

Our RPC method middleware / handler types are a hot mess. This makes
them slightly less odious to work with in the extension, although they
will still have to be fixed upstream.

rekmarks committed:
refactor: Revert TypeScript conversion

Reverts the TypeScript conversion of method middleware files to make
review easier.

legobeat committed:
Merge branch 'develop' into remove-unused-rpc-handler-hook

rekmarks commented on app/scripts/lib/rpc-method-middleware/handlers/index.ts:
```diff
@@ -17,15 +16,7 @@ import mmiSetAccountAndNetwork from './institutional/mmi-set-account-and-network
 import mmiOpenAddHardwareWallet from './institutional/mmi-open-add-hardware-wallet';
 ///: END:ONLY_INCLUDE_IF
 
-type MessageType = (typeof MESSAGE_TYPE)[keyof typeof MESSAGE_TYPE];
-
-type HandlerInterface = {
-  methodNames: MessageType[];
-  implementation: unknown;
-  hookNames?: { [key: string]: boolean };
-};
-
-const handlers: HandlerInterface[] = [
```
`handlers` is self-declaring.

rekmarks committed:
test: Add more error test cases

rekmarks commented on jest.config.js:
```diff
@@ -10,7 +10,6 @@ module.exports = {
     '<rootDir>/app/scripts/controllers/preferences.js',
     '<rootDir>/app/scripts/flask/**/*.js',
     '<rootDir>/app/scripts/lib/**/*.(js|ts)',
-    '<rootDir>/app/scripts/lib/createRPCMethodTrackingMiddleware.js',
```
Obviously extraneous due to the line above it.

rekmarks committed:
Merge branch 'develop' into remove-unused-rpc-handler-hook

rekmarks committed:
Merge branch 'develop' into remove-unused-rpc-handler-hook

rekmarks committed:
Merge branch 'develop' into remove-unused-rpc-handler-hook

rekmarks committed:
Merge branch 'develop' into remove-unused-rpc-handler-hook

rekmarks committed:
Merge branch 'develop' into remove-unused-rpc-handler-hook

legobeat commented:
It does seem like this could be handled more consistently at the type level by TypeScript, but I'm assuming the issue is that there is still a lot of unmigrated js code that will benefit from this in the meantime?

rekmarks commented:
@legobeat there's definitely unmigrated JS code, but also the types for the method middleware and its handler + hooks pattern are currently broken and implemented in 3 different repositories. <details><summary>🙃</summary>We're all trying to find the guy who did this...</details>

I may try to wrangle the types once this is in.

adonesky1 reviewed (approved):
Makes sense to me!

rekmarks committed:
Merge branch 'develop' into remove-unused-rpc-handler-hook

adonesky1 committed:
Merge branch 'develop' into remove-unused-rpc-handler-hook
